src/shell/shell.cpp

```diff
@@ -452,6 +453,49 @@ void DOS_Shell::SyntaxError()
 	WriteOut(MSG_Get("SHELL_SYNTAX_ERROR"));
 }
 
+static std_fs::path get_cmd_history_path()
+{
+	return get_platform_config_dir() / "cmd_history.txt";
+}
+
+void DOS_Shell::ReadCommandHistory()
+{
+	std::ifstream history_file(get_cmd_history_path());
+	if (history_file) {
+		std::string line;
+		while (getline(history_file, line)) {
+			if (!line.empty()) {
+				history.push_back(line);
+			}
+		}
+	}
+}
+
+void DOS_Shell::WriteCommandHistory()
+{
+	std_fs::path history_path = get_cmd_history_path();
+	std::ofstream history_file(history_path);
+	if (!history_file) {
+		LOG_WARNING("SHELL: Unable to write command history to file: %s", history_path.string().c_str());
+		return;
+	}
+	std::vector<std::string> trimmed_history;
+	trimmed_history.reserve(history.size());
+	for (std::string& str : history) {
+		if (!str.empty()) {
+			trimmed_history.push_back(str);
+		}
+	}
+	size_t start = 0;
+	size_t size = trimmed_history.size();
+	if (size > CMD_MAX_HISTORY_FILE_LINES) {
+		start = size - CMD_MAX_HISTORY_FILE_LINES;
+	}
```

Comment:
minor: Just use `std::max` or `clamp` from `math_utils.h`?

Reply:
Sure.  I had to switch to signed int for that to work but I don't think this vector will ever be big enough to cause an overflow.